go-controller/pkg/networkmanager/network_controller.go

```diff
@@ -124,84 +227,297 @@ func (c *networkController) getAllNetworkStates() []*networkControllerState {
+		}
+	}
+	klog.Infof("%s: finished syncing all networks. Time taken: %s", c.name, time.Since(start))
+	return nil
+}
+
+func (c *networkController) syncRunningNetworks() error {
+	c.networkReconciler.Reconcile(types.DefaultNetworkName)
+	for _, network := range c.getAllNetworkStates() {
+		c.networkReconciler.Reconcile(network.controller.GetNetworkName())
+	}
+
+	return nil
+}
+
+// syncNetwork must be called with nm mutex locked
+func (c *networkController) syncNetwork(network string) error {
 	startTime := time.Now()
 	klog.V(5).Infof("%s: sync network %s", c.name, network)
 	defer func() {
 		klog.V(4).Infof("%s: finished syncing network %s, took %v", c.name, network, time.Since(startTime))
 	}()
 
+	have, stoppedAndDeleting := c.getReconcilableNetworkState(network)
 	want := c.getNetwork(network)
-	have := c.getNetworkState(network)
 
-	// we will dispose of the old network if deletion is in progress or if
-	// configuration changed
-	dispose := have != nil && (have.stoppedAndDeleting || !util.AreNetworksCompatible(have.controller, want))
+	compatible := util.AreNetworksCompatible(have, want)
 
+	// we will dispose of the old network if deletion is in progress or if
+	// non-reconcilable configuration changed
+	dispose := stoppedAndDeleting || !compatible
 	if dispose {
-		if !have.stoppedAndDeleting {
-			have.controller.Stop()
-		}
-		have.stoppedAndDeleting = true
-		err := have.controller.Cleanup()
+		err := c.deleteNetwork(network)
 		if err != nil {
-			return fmt.Errorf("%s: failed to cleanup network %s: %w", c.name, network, err)
+			return err
 		}
-		c.setNetworkState(network, nil)
+		have = nil
 	}
 
-	// no network needed so nothing to do
-	if want == nil {
+	// fetch other relevant network information
+	err := c.gatherNetwork(want)
+	if err != nil {
+		return fmt.Errorf("failed to fetch other network information for network %s: %w", network, err)
+	}
+
+	ensureNetwork := !compatible || util.DoesNetworkNeedReconciliation(have, want)
+	if !ensureNetwork {
+		// no network changes
 		return nil
 	}
 
-	// we didn't dispose of current controller, so this might just be an update of the network NADs
-	if have != nil && !dispose {
-		have.controller.SetNADs(want.GetNADs()...)
+	// ensure the network controller
+	err = c.ensureNetwork(want)
+	if err != nil {
+		return fmt.Errorf("%s: failed to ensure network %s: %w", c.name, network, err)
+	}
+
+	return nil
+}
+
+func (c *networkController) ensureNetwork(network util.MutableNetInfo) error {
+	if network == nil {
 		return nil
 	}
 
-	// setup & start the new network controller
-	nc, err := c.cm.NewNetworkController(want)
+	networkName := network.GetNetworkName()
+	reconcilable, _ := c.getReconcilableNetworkState(networkName)
+
+	// this might just be an update of reconcilable network configuration
+	if reconcilable != nil {
+		err := reconcilable.Reconcile(network)
+		if err != nil {
+			return fmt.Errorf("failed to reconcile network %s: %w", networkName, err)
+		}
+		return nil
```

Comment:
so it seems like we are loading this function with 2 abilities:
1. if something is reconcilable then we reconcile it and return out
2. otherwise we start a new controller

so does "reconcilable" mean that a controller already exists? If so, then why dont we just call it that?

Reply:
I can do that. For me it helps understand for birds eye view that the purpose of that is to be reconciled. But I don't have a problem naming that `nc` or `controller`